File: qubo/frontend/src/types/analytics.ts
export interface TopicPerformance {
  topic_id: string;
  topic_name: string;
  difficulty_level: string | null;
  score_percentage: number | null;
  sessions_count: number;
  last_updated: string | null;
}

export interface QuizScorePoint {
  score: number;
  attempted_at: string;
}

export interface ExamPrediction {
  id: string;
  subject_id: string;
  subject_name: string;
  predicted_score: number;
  risk_level: 'low' | 'medium' | 'high';
  threshold: number;
  is_warning: boolean;
  basis_attempt_count: number;
  generated_at: string;
}

export interface SubjectAnalytics {
  id: string;
  subject_name: string;
  category: string | null;
  overall_mastery: number | null;
  topics_total: number;
  topics_measured: number;
  study_minutes: number;
  study_sessions: number;
  quizzes_completed: number;
  learning_velocity: number | null;
  topic_performance: TopicPerformance[];
  recent_quiz_scores: QuizScorePoint[];
  latest_prediction: ExamPrediction | null;
}

export interface StudySessionCreate {
  subject_id: string;
  topic_name: string;
  duration_minutes: number;
  pomodoro_cycles: number;
  session_date?: string;
  notes?: string;
}

export interface StudySession extends StudySessionCreate {
  id: string;
  subject_name: string;
  topic_id: string;
  session_date: string;
}

export interface ReviewSchedule {
  id: string;
  subject_id: string;
  subject_name: string;
  topic_id: string;
  topic_name: string;
  ease_factor: number;
  interval_days: number;
  repetitions: number;
  next_review_date: string;
  last_reviewed_date: string | null;
  last_score: number | null;
  is_due: boolean;
  quiz_id: string | null;
}

export interface AnalyticsFilters {
  subject_id?: string;
  topic_id?: string;
  date_from?: string;
  date_to?: string;
}

export interface EducatorStudentSummary {
  id: string;
  username: string;
  full_name: string;
  school: string | null;
  form_level: string | null;
  target_grade: string | null;
  profile_picture_url: string | null;
  subject_count: number;
  study_minutes: number;
  quizzes_completed: number;
  average_mastery: number | null;
  latest_prediction: number | null;
  at_risk: boolean;
  subjects: SubjectAnalytics[];
}

export interface EducatorDashboard {
  summary: {
    linked_students: number;
    students_at_risk: number;
    average_prediction: number | null;
    total_study_minutes: number;
    completed_quizzes: number;
  };
  students: EducatorStudentSummary[];
}

export interface StudyPlanRecommendation {
  id: string;
  student_id: string;
  recommendation_type: string;
  subject_id: string | null;
  subject_name: string | null;
  topic_id: string | null;
  topic_name: string | null;
  recommendation_text: string;
  priority_level: number;
  resource_link: string | null;
  is_accepted: boolean;
  created_at: string;
}

export interface StudyPlanResponse {
  message: string;
  recommendations: StudyPlanRecommendation[];
}

File: qubo/backend/app/schemas/analytics.py
from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, Field, field_validator


class TopicPerformanceItem(BaseModel):
    topic_id: str
    topic_name: str
    difficulty_level: Optional[str] = None
    score_percentage: Optional[float] = None
    sessions_count: int = 0
    last_updated: Optional[datetime] = None


class QuizScorePoint(BaseModel):
    score: float
    attempted_at: datetime


class PredictionItem(BaseModel):
    id: str
    subject_id: str
    subject_name: str
    predicted_score: float
    risk_level: str
    threshold: float
    is_warning: bool
    basis_attempt_count: int
    generated_at: datetime


class SubjectAnalyticsItem(BaseModel):
    id: str
    subject_name: str
    category: Optional[str] = None
    overall_mastery: Optional[float] = None
    topics_total: int = 0
    topics_measured: int = 0
    study_minutes: int = 0
    study_sessions: int = 0
    quizzes_completed: int = 0
    learning_velocity: Optional[float] = None
    topic_performance: List[TopicPerformanceItem]
    recent_quiz_scores: List[QuizScorePoint]
    latest_prediction: Optional[PredictionItem] = None


class StudySessionCreate(BaseModel):
    subject_id: str
    topic_name: str = Field(..., min_length=2, max_length=120)
    duration_minutes: int = Field(..., ge=1, le=720)
    pomodoro_cycles: int = Field(default=0, ge=0, le=30)
    session_date: Optional[datetime] = None
    notes: Optional[str] = Field(default=None, max_length=500)

    @field_validator("topic_name")
    @classmethod
    def clean_topic_name(cls, value: str) -> str:
        cleaned = " ".join(value.split())
        if len(cleaned) < 2:
            raise ValueError("Topic name must contain at least 2 characters")
        return cleaned


class StudySessionItem(BaseModel):
    id: str
    subject_id: str
    subject_name: str
    topic_id: str
    topic_name: str
    duration_minutes: int
    pomodoro_cycles: int
    session_date: datetime
    notes: Optional[str] = None


class ReviewScheduleItem(BaseModel):
    id: str
    subject_id: str
    subject_name: str
    topic_id: str
    topic_name: str
    ease_factor: float
    interval_days: int
    repetitions: int
    next_review_date: str
    last_reviewed_date: Optional[str] = None
    last_score: Optional[float] = None
    is_due: bool
    quiz_id: Optional[str] = None


class PredictionSettings(BaseModel):
    threshold: float = Field(..., ge=0, le=100)


class PredictionSettingsResponse(BaseModel):
    threshold: float


class StudentLinkRequest(BaseModel):
    username: str = Field(..., min_length=3, max_length=50)

    @field_validator("username")
    @classmethod
    def clean_username(cls, value: str) -> str:
        return value.strip()


class EducatorStudentSummary(BaseModel):
    id: str
    username: str
    full_name: str
    school: Optional[str] = None
    form_level: Optional[str] = None
    target_grade: Optional[str] = None
    profile_picture_url: Optional[str] = None
    subject_count: int = 0
    study_minutes: int = 0
    quizzes_completed: int = 0
    average_mastery: Optional[float] = None
    latest_prediction: Optional[float] = None
    at_risk: bool = False
    subjects: List[SubjectAnalyticsItem]


class EducatorClassSummary(BaseModel):
    linked_students: int
    students_at_risk: int
    average_prediction: Optional[float] = None
    total_study_minutes: int
    completed_quizzes: int


class EducatorDashboardResponse(BaseModel):
    summary: EducatorClassSummary
    students: List[EducatorStudentSummary]


class StudentLinkResponse(BaseModel):
    id: str
    message: str
